File: Basic/BehavioralPattern/src/org/uncle/lee/command/Invoker.java
package org.uncle.lee.command;

import java.util.ArrayList;
import java.util.List;

public class Invoker {
	private List<Command> commandList;
	
	public Invoker(){
		commandList = new ArrayList<Command>();
	}
	
	public boolean addCommand(Command command){
		return commandList.add(command);
	}
	
	public boolean removeCommand(Command command){
		return commandList.remove(command);
	}
	
	public void invoke(){
		for(Command command : commandList){
			command.execute();
		}
	}
}

File: Section5/src/behavioral/command/Invoker.kt
package behavioral.command

class Invoker {
    private val commandList : MutableList<Command> = mutableListOf()

    public fun addCommand(command : Command) {
        if (!this.commandList.contains(command)) {
            this.commandList.add(command)
        }
    }

    public fun removeCommand(command : Command) {
        if (this.commandList.contains(command)) {
            this.commandList.remove(command)
        }
    }

    public fun call() {
        this.commandList.forEach {
            it.execute()
        }
    }
}